### PR TITLE
Renderer tests

### PR DESCRIPTION
**Pull request by fywu85**
## Pull request information

- **Status**: ready to merge
- **Kind of changes**: new feature / documentation
- **Related PR or issue**: #581 

## Description

+ Increase test coverage of pyglet-renderer from ~30% to ~100%. 
+ Add attributes to pyglet renderer docstrings.
+ Rename internal methods of the renderer class for better readability.
+ Add `xvfb` service to `travis.yml` (`xvfb-run` will not work because of incomprehensible incompatibility with `ray`), along with a few minor fixes.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,7 +7,10 @@ python:
 
 os: linux
 
-dist: trusty
+dist: xenial
```
Comment (AboudyKreidieh) [inline .travis.yml]: why do you want to switch from 14.04? Do the changes not allow us to support it anymore?

File: flow/renderer/pyglet_renderer.py [hunk 2]
```diff
@@ -55,8 +75,37 @@ def __init__(self, network, mode,
...
-                 pxpm=2):
-        """Instantiate a pyglet renderer class."""
+                 pxpm=2,
+                 alpha=1.0):
+        """Initialize Pyglet Renderer.
+
+        Parameters
+        ----------
+        network : list
+            A list of road network polygons
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]: can you be more specific? maybe just by specifying `list of ...`

File: flow/renderer/pyglet_renderer.py [hunk 3]
```diff
@@ -55,8 +75,37 @@ def __init__(self, network, mode,
...
+            Specify whether to save rendering data to disk
+        path : str
+            Specify where to store the rendering data
+        sight_radius : int
+            Set the radius of observation for RL vehicles (meter)
+        show_radius : bool
+            Specify whether to render the radius of RL observation
+        pxpm : int
+            Specify rendering resolution (pixel / meter)
+        alpha: int
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]:
```suggestion
        alpha : int
```

File: flow/renderer/pyglet_renderer.py [hunk 4]
```diff
@@ -55,8 +75,37 @@ def __init__(self, network, mode,
...
+            Specify opacity of the alpha channel.
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]: what is the alpha channel?

File: flow/renderer/pyglet_renderer.py [hunk 5]
```diff
@@ -163,17 +217,10 @@ def render(self,
             Set the radius of observation for RL vehicles (meter)
         show_radius : bool
             Specify whether to render the radius of RL observation
+        sleep: float
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]:
```suggestion
        sleep : float
```

File: flow/renderer/pyglet_renderer.py [hunk 6]
```diff
@@ -222,19 +298,35 @@ def render(self,
...
         self.window.flip()
 
-        if "gray" in self.mode:
-            _frame = cv2.cvtColor(self.frame, cv2.COLOR_BGR2GRAY)
-        else:
-            _frame = self.frame
-        if save_render:
+        # import matplotlib.pyplot as plt
+        # plt.imshow(self.frame)
+        # plt.show()
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]: maybe delete the commented stuff?

File: flow/renderer/pyglet_renderer.py [hunk 7]
```diff
@@ -391,7 +479,18 @@ def _add_vehicle_poly_circle(self, center, radius, color):
...
+        self.vehicle_batch.add_indexed(
+            pxpm, pyglet.gl.GL_LINE_LOOP, group, index,
+            ("v2f", vertex_list), ("c4B", vertex_color))
+
+    def _truncate_colormap(self,
+                           cmap,
+                           minval=0.25,
+                           maxval=0.75,
+                           n=100):
+        """Truncate a matplotlib colormap."""
```
Comment (AboudyKreidieh) [inline flow/renderer/pyglet_renderer.py]: Maybe add what each parameter is? most of them are clear, but `n` is confusing. Alternatively, you can just give `n` a clearer name and then describing the parameters shouldn't be necessary

File: tests/fast_tests/test_pyglet_renderer.py [hunk 8]
```diff
@@ -1,85 +1,227 @@
...
 import os
 import unittest
-
-os.environ['TEST_FLAG'] = 'True'
+import ctypes
 
 
 class TestPygletRenderer(unittest.TestCase):
     """Tests pyglet_renderer"""
+    def setUp(self):
```
Comment (AboudyKreidieh) [inline tests/fast_tests/test_pyglet_renderer.py]:
```suggestion

    def setUp(self):
```

Comment (AboudyKreidieh) [reply tests/fast_tests/test_pyglet_renderer.py]: this suggestion is so the changes don't cause failures when something else is merged

File: .travis.yml (hunk 1 above)
Comment (fywu85) [reply .travis.yml]: `xvfb` service is only available on 16.04 and above so I had to bump the version up. Meanwhile, 14.04 came to its end of life cycle this April so it is probably a good idea for us to move on to something newer, assuming this does not break anything in the process.

File: flow/renderer/pyglet_renderer.py (hunk 4 above)
Comment (fywu85) [reply flow/renderer/pyglet_renderer.py]: Alpha channel controls transparency of each visual element. For more info, check https://en.wikipedia.org/wiki/Alpha_compositing

File: flow/renderer/pyglet_renderer.py (hunk 2 above)
Comment (fywu85) [reply flow/renderer/pyglet_renderer.py]: Added detailed comments.

File: flow/renderer/pyglet_renderer.py (hunk 6 above)
Comment (fywu85) [reply flow/renderer/pyglet_renderer.py]: Stuff deleted.

File: flow/renderer/pyglet_renderer.py (hunk 7 above)
Comment (fywu85) [reply flow/renderer/pyglet_renderer.py]: Docs added.

**Comment by fywu85** [comment]
Thanks. Should address everything we have so far. Let me know if you have additional inputs.

**Pull request merged**